pkg/proc/gdbserial/undo.go

```diff
@@ -2,19 +2,318 @@ package gdbserial
+		return "", err
+	}
+	recording_ids, err := p.conn.undoCmd("get_recording_ids")
+	if err != nil {
+		return "", err
+	}
+	uuids := strings.Split(recording_ids, ";")
+
+	// This directory stores sessions.
+	xdg_data_dir, present := os.LookupEnv("XDG_DATA_HOME")
+	if !present {
+		xdg_data_dir = filepath.Join(user.HomeDir, ".local", "share")
+	}
+	undo_sessions_dir := filepath.Join(xdg_data_dir, "undo", "sessions")
+
+	err = os.MkdirAll(undo_sessions_dir, 0755)
+	if err != nil {
+		return "", err
+	}
+
+	file := filepath.Join(undo_sessions_dir, string(uuids[1])+".json")
+
+	return file, nil
+}
+
+// Load the UDB session file (if it exists) for the current recording.
+func (uc *undoSession) load(p *gdbProcess) error {
+	path, err := getSessionPath(p)
+	if err != nil {
+		return err
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	decoder := json.NewDecoder(file)
+	decoder.DisallowUnknownFields()
+
+	var s session
+	err = decoder.Decode(&s)
+	if err != nil {
+		return err
+	}
+
+	// Clear out the session data.
+	*uc = *newUndoSession()
+
+	// Translate the loaded Undo bookmarks into Delve checkpoints.
+	for name, position := range s.Bookmarks {
+		cpid := uc.checkpointLastId
+		uc.checkpointLastId++
+		uc.checkpoints[cpid] = proc.Checkpoint{
+			ID:    cpid,
+			When:  fmt.Sprintf("%x,%x", position.Bbcount, position.Pc),
+			Where: name,
+		}
+	}
+
+	return err
+}
+
+// Save the session file for the current recording.
+func (uc *undoSession) save(p *gdbProcess) error {
+	// Translate Delve checkpoints into Undo bookmarks.
+	var s session
+	s.Bookmarks = make(map[string]bookmarkTime)
+
+	// Local copy of the checkopints.
+	var checkpoints []proc.Checkpoint
+	for _, cp := range uc.checkpoints {
+		checkpoints = append(checkpoints, cp)
+	}
+	// Sort the checkpoints by descending note length - this is to avoid adding a suffix to a
+	// entries that we've already added with a suffix to. e.g. if we've previously saved this
+	// session with a duplicated checkpoint note called "test" will have extended one to
+	// "test-0". If the user adds "test" again, we want to rename that to "test-1" rather than
+	// creating a "test-0-0".
+	sort.Slice(checkpoints, func(i, j int) bool {
+		// This is a "Less" function that sorts in descending order of string length.
+		return len(checkpoints[i].Where) > len(checkpoints[j].Where)
+	})
+
+	// Iterate through sorted checkpoints to eliminate duplicates.
+	for _, cp := range checkpoints {
+		// Ensure that notes are made unique before saving - UDB expects bookmark names to
+		// be unique.
+		base_name := cp.Where
+		name := base_name
+		for i := 0; s.Bookmarks[name] != (bookmarkTime{}); i++ {
+			name = fmt.Sprintf("%s-%d", base_name, i)
+		}
+
+		var time bookmarkTime
+		_, err := fmt.Sscanf(cp.When, "%x,%x", &time.Bbcount, &time.Pc)
+		if err != nil {
+			return err
+		}
+		s.Bookmarks[name] = time
+	}
+
+	path, err := getSessionPath(p)
+	if err != nil {
+		return err
+	}
```

Comment:
Might be worth doing this first — if there's no loaded recording then we can skip some work.